src/osd/OSD.cc

```diff
@@ -1300,7 +1300,7 @@ void OSDService::got_stop_ack()
 MOSDMap *OSDService::build_incremental_map_msg(epoch_t since, epoch_t to,
                                                OSDSuperblock& sblock)
 {
-  MOSDMap *m = new MOSDMap(monc->get_fsid());
+  MOSDMap *m = new MOSDMap(monc->get_fsid(), osdmap->get_features(entity_name_t::TYPE_OSD, NULL));
```

Comment:
wrap at 79 chars.

Reply:
this is the wrong method.  get_features() tells you what client featuers are required given what's in teh osdmap, not how the osdmap is/should be encoded.

I think we actaully needt o take this block from OSDMonitor::encode_pending() and move it into an OSDMap method, get_encode_features():
```
    // determine appropriate features
    if (tmp.require_osd_release < CEPH_RELEASE_LUMINOUS) {
      dout(10) << __func__ << " encoding without feature SERVER_LUMINOUS"
	       << dendl;
      features &= ~CEPH_FEATURE_SERVER_LUMINOUS;
    }
    if (tmp.require_osd_release < CEPH_RELEASE_KRAKEN) {
      dout(10) << __func__ << " encoding without feature SERVER_KRAKEN | "
	       << "MSG_ADDR2" << dendl;
      features &= ~(CEPH_FEATURE_SERVER_KRAKEN |
		    CEPH_FEATURE_MSG_ADDR2);
    }
    if (tmp.require_osd_release < CEPH_RELEASE_JEWEL) {
      dout(10) << __func__ << " encoding without feature SERVER_JEWEL" << dendl;
      features &= ~CEPH_FEATURE_SERVER_JEWEL;
    }
    if (tmp.require_osd_release < CEPH_RELEASE_MIMIC) {
      dout(10) << __func__ << " encoding without feature SERVER_MIMIC" << dendl;
      features &= ~(CEPH_FEATURE_SERVER_MIMIC);
    }
    dout(10) << __func__ << " encoding full map with " << features << dendl;
```
And the OSDMonitor code should just get an assert to make sure whatever the OSDMap method returns is a subset of the mon quorum features.

Reply:
but this is not a self contained piece, it based on 
```
  // features for osdmap and its incremental
  uint64_t features = mon->get_quorum_con_features();
```
so OSDMap itself cannot come out with a feature-set, but OSDMonitor find out proper features and use that to encode the map.

Reply:
I think the issue is , if we get something from bl cache or from disk, do we have anyway to tell how the BL was encoded before fully decode it? 

the paxosservice::get_version in OSDMonitor looks like have same issue.

Reply:
not easily, no.  i added a comment in get_version[_full] to that effect. basically i think it doesn't matter.. we'll just (worst case) cache the same encoded osdmap under the latest mon quorum features and also the encoded osdmap's features.